tests/test_local_code.py

```diff
@@ -38,66 +35,195 @@
-        assert len(extra_paths) == 1
+        zip_file, zip_python_paths, cwd = result
         with zipfile.ZipFile(zip_file) as zip:
-            all_paths = zip_python_paths + extra_paths
+            all_paths = zip_python_paths + [cwd]
             for file in zip.filelist:
                 ext = os.path.splitext(file.filename)[1]
                 assert ext == ".py"
                 assert any(zip_path in file.filename for zip_path in all_paths)
+
+
+@pytest.mark.parametrize(
+    "available_files,working_directory,working_directory_globs,python_paths,"
+    "expected_files_in_zip,expected_python_paths_in_zip,"
+    "expected_working_directory_in_zip",
+    [
+        # root is not on python path, but subdirectories are. Python files outside of
+        # the subdirectories should be ignored. Also, non-.py files are ignored
+        (
+            [
+                "root/a/b/x1.py",
+                "root/a/c/x2.py",
+                "root/a/x3.py",
+                "root/x4.py",
+                "root/a/x5.txt",
+                "root/a/b/x6.txt",
+            ],
+            "root",
+            (),
+            ["root/a/b", "root/a/c"],
+            ["root/a/b/x1.py", "root/a/c/x2.py"],
+            ["root/a/b", "root/a/c"],
+            "root",
+        ),
+        # simple case of include **/*.txt files
+        (
+            [
+                "root/a/x1.py",
+                "root/a/x2.txt",
+                "root/b/x3.py",
+                "root/b/x4.txt",
+                "root/x5.txt",
+            ],
+            "root",
+            ["**/*.txt"],
+            ["root"],
+            [
+                "root/a/x1.py",
+                "root/a/x2.txt",
+                "root/b/x3.py",
+                "root/b/x4.txt",
+                "root/x5.txt",
+            ],
+            ["root"],
+            "root",
+        ),
+        # same as above, but only include txt files in b
+        (
+            [
+                "root/a/x1.py",
+                "root/a/x2.txt",
+                "root/b/x3.py",
+                "root/b/x4.txt",
+                "root/x5.txt",
+            ],
+            "root",
+            ["b/**/*.txt"],
+            ["root"],
+            ["root/a/x1.py", "root/b/x3.py", "root/b/x4.txt"],
+            ["root"],
+            "root",
+        ),
+        # reference files outside of root
+        (
+            ["root/x1.py", "root/x2.txt", "other_root/x3.txt"],
+            "root",
+            ["../other_root/**/*.txt"],
+            ["root"],
+            ["root/x1.py", "other_root/x3.txt"],
+            ["root"],
+            "root",
+        ),
+        # reference files two folders outside of root
+        (
+            ["nested/root/x1.py", "nested/root/x2.txt", "other_root/x3.txt"],
+            "nested/root",
+            ["../../other_root/**/*.txt"],
+            ["nested/root"],
+            # TODO this is actually busted! this needs to be nested/root/x1.py in order
```

Comment:
Okay, so this basically works the way we discussed, except you do "../../foo" (a single "../foo" works, kind of by "accident"). I'm tempted to just leave this bug in for now and deal with it later when it becomes an issue?

Reply:
Okay, I just made it throw an exception if you try to do this. We can still add support for it later if it seems useful